### PR TITLE
Add configuration context to dataset

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This adds a globally propagated configuration context to Dataset, where configs can be set on the driver and workers can retrieve them via singleton pattern. This avoids redundant plumbing of configuration / resources like the block owner through developer / public APIs, which may make API stability difficult to maintain.

### CODE REVIEW

**Review by jjyao — APPROVED** [review]
Wait for clark's approval as well.

File: python/ray/data/context.py
Comment (ericl) [reply python/ray/data/context.py:24]: It will be in the next PR (can leave it for now).

File: python/ray/data/context.py
Comment (ericl) [reply python/ray/data/context.py]: Yeah making all Dataset APIs thread safe is a good idea, fixed.

**Review by scv119 — APPROVED** [review]

**Pull request merged**
